### PR TITLE
[4.0] Filter by parent category

### PR DESCRIPTION
**Pull request by brianteeman**
On a complex site with a lot of categories it can be very difficult to sort the categories using drag and drop because you cannot get the relevant tree on the same page.

This adds a new filter to com_category that allows you to select the parent category. It's particularly useful when used with the max levels filter.

Testing is easiest if you have the test sample data installed as that has a lot of categories

### Example with no filter
image

### Example with category filter
image

### Example with category and level filter
image

### Example with TWO category filters and level filter
image

### CODE REVIEW

**Comment by Quy** [comment]
Will this fix #22076?

**Comment by brianteeman** [comment]
It would help but its not the same thing

File: administrator/components/com_categories/Model/CategoriesModel.php [hunk 1]
```diff
@@ -234,11 +237,37 @@ protected function getListQuery()
...
+			$categoryId = $categoryId ? array($categoryId) : array();
+		}
+
+		// Case: Using both categories filter and by level filter
+		if (count($categoryId))
+		{
+			$categoryTable = Table::getInstance('Category', 'JTable');
+			$subCatItemsWhere = array();
+			foreach ($categoryId as $filter_catid)
+
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoriesModel.php]:
```suggestion
```

File: administrator/components/com_categories/Model/CategoriesModel.php [hunk 2]
```diff
@@ -234,11 +237,37 @@ protected function getListQuery()
...
+		{
+			$categoryId = $categoryId ? array($categoryId) : array();
+		}
+
+		// Case: Using both categories filter and by level filter
+		if (count($categoryId))
+		{
+			$categoryTable = Table::getInstance('Category', 'JTable');
+			$subCatItemsWhere = array();
+			foreach ($categoryId as $filter_catid)
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoriesModel.php]:
```suggestion

			foreach ($categoryId as $filter_catid)
```

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 38241f772cf2edc00b87b836d6b29c54a38a1345<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27486">issues.joomla.org/tracker/joomla-cms/27486</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 16949486f6ecfb86d547a47fa9bd1a7d470ecac9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27486">issues.joomla.org/tracker/joomla-cms/27486</a>.</sub>

File: administrator/components/com_categories/Model/CategoriesModel.php [hunk 3]
```diff
@@ -17,7 +17,9 @@
 use Joomla\CMS\Language\Associations;
 use Joomla\CMS\MVC\Factory\MVCFactoryInterface;
 use Joomla\CMS\MVC\Model\ListModel;
+use Joomla\CMS\Table\Table;
 use Joomla\Database\ParameterType;
+use Joomla\Utilities\ArrayHelper;
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoriesModel.php]:
```suggestion
```

File: administrator/components/com_categories/Model/CategoriesModel.php [hunk 4]
```diff
@@ -234,11 +237,37 @@ protected function getListQuery()
...
+
+			$query->where('(' . implode(' OR ', $subCatItemsWhere) . ')');
+		}
+
+		// Case: Using only the by level filter
+		elseif ($level)
 		{
-			$query->where($db->quoteName('a.level') . ' <= :level')
-				->bind(':level', $level, ParameterType::INTEGER);
+			$query->where('a.level <= ' . (int) $level);
```
Comment (Quy) [inline administrator/components/com_categories/Model/CategoriesModel.php]: Revert line 240-241.

**Comment by sakiss** [comment]
I have tested this item :white_check_mark: successfully on 16949486f6ecfb86d547a47fa9bd1a7d470ecac9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27486">issues.joomla.org/tracker/joomla-cms/27486</a>.</sub>

**Comment by brianteeman** [comment]
drone errors are the usual flaky test and nothing to do with this pr

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27486">issues.joomla.org/tracker/joomla-cms/27486</a>.</sub>

**Comment by richard67** [comment]
@brianteeman I've just tested it before I wanted to merge but I've noticed something:

In categories list view for com_content everything works fine, but in categories list view for com_contact only categories of com_content are available for selection in the new categories filter field.

So it seems I always can select only com_content categories, regardless in which component's categories view I am.

I used testing sample data on current 4.0-dev.

Can you check and confirm or let me know if I'm doing something wrong?

And what is desired behavior of the new filter: Always be able to select from all categories, regardless of component (not nice), or should it be aware of the component and only offer suitable categories for the particular component for selection (perfect)?

**Comment by infograf768** [comment]
I confirm the issue. Setting back to pending.



Note: I also have issue merging this PR in my local repo. Had to pick up the file from PR branch.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27486">issues.joomla.org/tracker/joomla-cms/27486</a>.</sub>

**Comment by infograf768** [comment]
I guess the issue may be in the xml where the filter is limited to com_content
```
<field
			name="category_id"
			type="category"
			label="JOPTION_SELECT_CATEGORY"
			multiple="true"
			extension="com_content". /// Limited to com_content
			layout="joomla.form.field.list-fancy-select"
			hint="JOPTION_SELECT_CATEGORY"
			onchange="this.form.submit();"
			published="0,1,2"
		/>
```

But it looks also that the branch may not be up to date.

**Comment by infograf768** [comment]
I propose to make the extension dynamic
```
		<field
			name="category_id"
			type="category"
			label="JOPTION_SELECT_CATEGORY"
			multiple="true"
			extension="dynamic"
			layout="joomla.form.field.list-fancy-select"
			hint="JOPTION_SELECT_CATEGORY"
			onchange="this.form.submit();"
			published="0,1,2"
		/>
```
and 
```
diff --git a/administrator/components/com_categories/src/View/Categories/HtmlView.php b/administrator/components/com_categories/src/View/Categories/HtmlView.php
index 2f1a431..275c322 100644
--- a/administrator/components/com_categories/src/View/Categories/HtmlView.php
+++ b/administrator/components/com_categories/src/View/Categories/HtmlView.php
@@ -127,4 +127,7 @@
 		}
 
+		$extensionName = $this->escape($this->state->get('filter.extension'));
+		$this->filterForm->setFieldAttribute('category_id', 'extension', $extensionName, 'filter');
+
 		return parent::display($tpl);
 	
```

Screen Shot 2020-02-26 at 09 33 18

**Comment by SharkyKZ** [comment]
@infograf768 Set the attribute in the model.

@brianteeman `getStoreId()` method in the model also needs to be updated.

**Comment by richard67** [comment]
@brianteeman Why closing? Will you make a new PR instead?

**Pull request closed without merging**